presto-orc/src/main/java/com/facebook/presto/orc/writer/LongDictionaryColumnWriter.java

```diff
@@ -0,0 +1,300 @@
+            directValues = Arrays.ensureCapacity(directValues, positionCount);
+            directNulls = Arrays.ensureCapacity(directNulls, positionCount);
+            for (int j = 0; j < positionCount; j++) {
+                if (segment[j] != UNUSED_VALUE) {
+                    directValues[j] = dictionary.getValue(segment[j]);
+                    directNulls[j] = false;
+                }
+                else {
+                    directNulls[j] = true;
+                }
+            }
+
+            LongArrayBlock longArrayBlock = new LongArrayBlock(positionCount, Optional.of(directNulls), directValues);
+            directColumnWriter.writeBlock(longArrayBlock);
+            if (directColumnWriter.getBufferedBytes() > maxDirectBytes) {
+                return false;
+            }
+            rowGroupValueCount -= positionCount;
+        }
+        checkState(rowGroupValueCount == 0);
+        return true;
+    }
+
+    @Override
+    protected ColumnEncoding getDictionaryColumnEncoding()
+    {
+        checkState(columnEncoding != null);
+        return columnEncoding;
+    }
+
+    @Override
+    protected BlockMetrics addBlockToDictionary(Block block, int rowGroupValueCount, IntBigArray rowGroupIndexes)
+    {
+        int nonNullValueCount = 0;
+        long rawBytes = 0;
+        for (int position = 0; position < block.getPositionCount(); position++, rowGroupValueCount++) {
+            if (block.isNull(position)) {
+                rowGroupIndexes.set(rowGroupValueCount, UNUSED_VALUE);
+            }
+            else {
+                long value = type.getLong(block, position);
+                int index = dictionary.addIfNotExists(value);
+                rowGroupIndexes.set(rowGroupValueCount, index);
+                statisticsBuilder.addValue(value);
+                rawBytes += typeSize;
+                nonNullValueCount++;
+            }
+        }
+        return new BlockMetrics(nonNullValueCount, rawBytes);
+    }
+
+    @Override
+    protected long getRetainedDictionaryBytes()
+    {
+        return dictionary.getRetainedBytes() + dictionaryDataStream.getRetainedBytes() + sizeOf(directValues) + sizeOf(directNulls);
+    }
+
+    @Override
+    protected int[] writeDictionary()
+    {
+        long[] elements = dictionary.elements();
+        for (int i = 0; i < dictionary.size(); i++) {
+            this.dictionaryDataStream.writeLong(elements[i]);
+        }
+
+        columnEncoding = new ColumnEncoding(DICTIONARY, dictionary.size());
+        return null;
+    }
+
+    @Override
+    protected void writePresentAndDataStreams(
+            int rowGroupValueCount,
+            IntBigArray rowGroupIndexes,
+            int[] originalDictionaryToSortedIndex,
+            PresentOutputStream presentStream,
+            LongOutputStream dataStream)
+    {
+        checkArgument(originalDictionaryToSortedIndex == null, "Unsupported originalDictionaryToSortedIndex");
```

Comment:
We don't usually pass nulls around. Make the abstract functions to pass Optionals and add javadoc to explain the reason to be optional.empty